lang/c/src/value-read-resolution.c

```diff
@@ -0,0 +1,628 @@
+        }
+        case AVRO_BYTES:
+        {
+            return avro_resolve_source(reader, source, dest, type_bytes);
+        }
+        case AVRO_DOUBLE:
+        {
+            return avro_resolve_source(reader, source, dest, type_double);
+        }
+        case AVRO_FLOAT:
+        {
+            return avro_resolve_source(reader, source, dest, type_float);
+        }
+        case AVRO_INT64:
+        {
+            return avro_resolve_source(reader, source, dest, type_long);
+        }
+        case AVRO_INT32:
+        {
+            return avro_resolve_source(reader, source, dest, type_int);
+        }
+        case AVRO_NULL:
+        {
+            return avro_resolve_source(reader, source, dest, type_null);
+        }
+            
+        case AVRO_STRING:
+        {
+            return avro_resolve_source(reader, source, dest, type_string);
+        }
+            
+        case AVRO_ARRAY:
+            return read_array_value_with_resolution(reader, source, dest);
+            
+        case AVRO_ENUM:
+        {
+            return avro_resolve_source(reader, source, dest, type_enum);
+        }
+            
+        case AVRO_FIXED:
+        {
+            return avro_resolve_source(reader, source, dest, type_fixed);
+        }
+            
+        case AVRO_MAP:
+            return read_map_value_with_resolution(reader, source, dest);
+            
+        case AVRO_RECORD:
+            return read_record_value_with_resolution(reader, source, dest);
+            
+        case AVRO_UNION:
+            return read_union_value_with_resolution(reader, source, dest);
+            
+        default:
+        {
+            avro_set_error("Unknown schema type");
+            return EINVAL;
+        }
+
+    }
+    
+    return 0;
+}
+
+int avro_value_read_with_resolution(avro_reader_t reader, avro_value_t *source, avro_value_t *dest)
+{
+    int  rval;
+    check(rval, avro_value_reset(dest));
+    avro_schema_t wschema = avro_value_get_schema(source);
+    avro_schema_t rschema = avro_value_get_schema(dest);
+    if(avro_schema_match(wschema, rschema) == 0)
+    {
+        avro_set_error("Schema resolution error!");
+        return EINVAL;
+    }
+    return read_value_with_resolution(reader, source, dest);
+}
+
+int
+read_union_value_with_resolution(avro_reader_t reader, avro_value_t *source, avro_value_t *dest)
+{
+    avro_schema_t wschema = avro_value_get_schema(source);
+    avro_schema_t rschema = avro_value_get_schema(dest);
+    
+    int wtype = avro_value_get_type(source);
+    int rtype = avro_value_get_type(dest);
+    if ( (rtype == AVRO_UNION) && (wtype == AVRO_UNION))
+    {
+        return resolve_unions(reader, source, dest);
+    }
+    else if ((rtype == AVRO_UNION) && (wtype != AVRO_UNION))
+    {
+        // reader is union, writer is not
+        int64_t  branch_count_dest;
+        branch_count_dest = avro_schema_union_size(rschema);
+        int i = 0;
```

Comment:
do you need this?

Reply:
Don't need the initialization (`= 0;`) part, or don't need the re-init in the for loop.  Probably redundant to have it both places.